aten/src/ATen/native/cuda/ReduceLogicKernel.cu

```diff
@@ -3,30 +3,33 @@
 #include <ATen/native/DispatchStub.h>
 #include <ATen/native/SharedReduceOps.h>
 #include <ATen/native/ReduceOps.h>
+#include <ATen/Dispatch.h>
 
 
 namespace at { namespace native {
 
 void and_kernel_cuda(TensorIterator& iter) {
-  AT_DISPATCH_ALL_TYPES_AND2(kHalf, kBool, iter.dtype(), "and_kernel", [&]() {
-    gpu_reduce_kernel<scalar_t, scalar_t>(
-        iter,
-        func_wrapper<scalar_t>([] GPU_LAMBDA(scalar_t a, scalar_t b) -> scalar_t {
-          return static_cast<scalar_t>(static_cast<bool>(a) && static_cast<bool>(b));
-        }),
-        static_cast<scalar_t>(true));
-  });
+  AT_DISPATCH_ALL_TYPES_AND3(
+      kHalf, kBFloat16, kBool, iter.common_dtype(), "and_cuda", [&]() {
+        gpu_reduce_kernel<scalar_t, bool>(
+            iter,
+            func_wrapper<bool>([] GPU_LAMBDA(bool a, bool b) -> bool {
+              return (a && b);
+            }),
+            true);
+      });
 }
 
 void or_kernel_cuda(TensorIterator& iter) {
-  AT_DISPATCH_ALL_TYPES_AND2(kHalf, kBool, iter.dtype(), "or_kernel", [&]() {
-    gpu_reduce_kernel<scalar_t, scalar_t>(
-        iter,
-        func_wrapper<scalar_t>([] GPU_LAMBDA(scalar_t a, scalar_t b) -> scalar_t {
-          return static_cast<scalar_t>(static_cast<bool>(a) || static_cast<bool>(b));
-        }),
-        static_cast<scalar_t>(false));
-  });
+  AT_DISPATCH_ALL_TYPES_AND3(
+      kHalf, kBFloat16, kBool, iter.common_dtype(), "or_cuda", [&]() {
+        gpu_reduce_kernel<scalar_t, bool>(
+            iter,
+            func_wrapper<bool>([] GPU_LAMBDA(bool a, bool b) -> bool {
```

Comment:
Two questions here:

- When gpu_reduce_kernel is templated it takes two arguments. What's the first one do, and what does the second one do?
- What would be the difference, if any, between taking the inputs in the lambda as scalar_ts and casting them to bool, like the older version did, and the current version that takes the inputs as bool?

Reply:
1. 
https://github.com/pytorch/pytorch/blob/10b490a3e0e502c35820fce96b2e6a82df96df44/aten/src/ATen/native/cuda/Reduce.cuh#L892-L894

First is the scalar type of data, and second is the scalar type of output.

2. Actually it makes sense to make it `(scalar_t a, scalar_t b) -> bool`, as then there won't be implicit cast at calling site. We can then use `static_cast` in the body, this also allows to use complex dtypes (as bool operator on them is explicit).
Thanks have updated the code and test!